feat(remap): Add `encode_logfmt` function

oktal opened the pull request:
Fixes #6972

This PR introduces a new `encode_logfmt` function, which is the opposite of `parse_logfmt`.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,227 @@
+
+fn format_logfmt_string(f: &mut fmt::Formatter<'_>, str: &str) -> fmt::Result {
+    match str.find(' ') {
+        Some(_) => write!(f, r#""{}""#, str),
```
Do we need to encode any `"`s appearing in the value?

oktal replied:
Good catch, I think we should escape strings the same way json strings are escaped

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,227 @@
+        &[
+            Example {
+                title: "encode object",
+                source: r#"encode_logfmt({"lvl": "info", "msg": "This is a message", "log_id": 12345})"#,
```
Does this handle nested objects and arrays? For example, I would expect:

```
{ "foo": { "bar": 1, "baz": [{"boop": 2}] }}
```

to encode as:

```
foo.bar=1 foo.baz.0.boop=2
```

oktal replied:
Currently, I implemented it using the `fmt::Display` trait implementation for `Value`.

I expect the result to be:

```
foo="{ "bar": 1, "baz": [{"boop": 2}] }"
```

Note that as you pointed out, quotes `"` are not properly escaped.

I don't think that nested keys are a common thing in `logfmt`.
What do you think ? Should I add a test-case ?

jszwedko replied:
👍 I think it would be nice to flatten them out. That's what I'm used to seeing for logfmt anyway.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,227 @@
+                                write!(f, "{}", field_formatter)?;
+                            }
+                        }
+                        _ => return Err(fmt::Error),
```
Could we wrap these errors to provide more context to users? For example, I might like to see output like: `Failed to encode as logfmt, index 4 has more than two elements` and `Failed to encode as logfmt, index 4 is an integer, not an array`.

We've commonly just passed around strings for errors in the VRL stdlib so you can wrap them by `format!("Oh no, an error occurred: {}", err)`ing as they bubble up.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,227 @@
+    fn parameters(&self) -> &'static [Parameter] {
+        &[Parameter {
+            keyword: "value",
+            kind: kind::OBJECT | kind::ARRAY,
```
If it simplifies things, I'm happy to just include the `OBJECT` form for now and leave array of key/value pairs for later.

oktal replied:
I added it because I'll need it for my own use case.

For example, if I have the given fields:
`{ "ts": 2021-02-04T17:01:02, "lvl": "info", "msg": "This is a log message", "trace_id": "abcd134fg" }`

I want my final encoded string to be:

`ts=2021-02-04T17:01:02 lvl=info msg="This is a log message" trace_id=abcd134fg`

jszwedko replied:
:+1: Sorry, I mean I'd be ok with not supporting the array form: `[["level", "debug"], ["msg": "hello"], ...]`.

oktal replied:
Ok so I've been thinking a little bit more about this.

Like I said, I added array of key-value pairs to guarantee ordering of my fields.
An other solution would be to add an optional array parameter to enforce the ordering like so:

```
encode_logfmt(object, [ "ts", "lvl", "msg" ])
```

What do you think ?

jszwedko replied:
Ah, I see, I missed that you wanted to specify an ordering. I think your suggestion makes sense. In that case, I would expect _only_ those fields to be serialized and that they be serialized in that order.

oktal replied:
Ah, I was expecting the opposite :D

Basically, for my own use case, I only want to guarantee the ordering of standard fields (`ts`, `lvl`, `msg`).

So maybe do we need a third parameter (maybe a boolean ?) to let the user decide whether the fields set should be "closed" or not.

jszwedko replied:
Ah! 😄 I think that's good as long as we document the behavior given that users can achieve the other behavior via passing in just the fields they want as the object to `encode_logfmt`.

binarylogic commented:
Nice! Thank you.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,227 @@
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value.type_def(state).infallible().bytes()
```
```suggestion
        self.value.type_def(state).fallible().bytes()
```

I think this function will actually always be fallible due to the way the encoding works for arrays where it can fail based on the input.

jszwedko reviewed (commented):
Thanks for this contribution @oktal !

I left some in-line comments. Additionally:

* could you add this function to the benchmarks: https://github.com/timberio/vector/blob/414f33f9aad407bab9b0664b8c1d65864b8c118a/lib/vrl/stdlib/benches/benches.rs ?
* could you add docs for this function? You can probably just copy and modify https://github.com/timberio/vector/blob/master/docs/reference/remap/functions/encode_json.cue

Let me know what you think!

oktal commented:
Hi @jszwedko so I think I addressed all your comments.

I removed the key-value pair syntax and added an optional `fields_ordering` parameter if one wants to guarantee the ordering of fields.

One thing I'm not really satisfied with is that the function is declared faillible since I'm using `std::fmt::Write` functions that can fail. How should we handle that ?

I'll also add docs

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,234 @@
+    fn encode_string(output: &mut String, str: &str) -> fmt::Result {
+        let needs_quotting = str.find(' ').is_some();
+        if needs_quotting {
+            output.write_char('"')?;
```
I think it is ok to use `expect()` here and avoid returning a `Result` given that writes to a `String` can never fail:

https://doc.rust-lang.org/src/alloc/string.rs.html#2418-2430

This would allow you to change the function to infallible.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,234 @@
+            None => vec![],
+            Some(expr) => match expr.resolve(ctx) {
+                Ok(val) => resolve_fields(&val),
+                Err(_) => vec![],
```
I think we should bubble up this error the way you did when resolving the value on 148.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,234 @@
+            },
+        };
+
+        let logfmt = match value {
```
I think you can use `try_object` here to avoid the error handling just below:

```rust
let value = value.try_object()?;
```

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,234 @@
+            .iter()
+            .filter_map(|v| match v {
+                Value::Bytes(bytes) => Some(String::from_utf8_lossy(&bytes)),
+                _ => None,
```
I think it'd be better to return an explicit error here if the array contains a value that is not a string.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,234 @@
+}
+
+fn resolve_fields(fields: &Value) -> Vec<String> {
+    match fields {
```
Similarly here, I think you should do `fields.try_array()?` and bubble up an error if `fields_ordering` is not an array.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,234 @@
+            output.write_char('"')?;
+        }
+
+        for c in str.chars() {
```
I am wondering if we should also be escaping things like newlines or carriage returns too. There was some related discussion to this in https://github.com/timberio/vector/pull/6739

jszwedko reviewed (commented):
Nice @oktal ! This is looking good.

I left some inline comments. Additionally, I think we still need support for flattening objects/arrays, if you agree, as I mentioned in https://github.com/timberio/vector/pull/6985#discussion_r606265314.

oktal commented:
Hey @jszwedko I updated this PR and added maps and arrays flattening like you suggested. I also added documentation for `encode_logfmt`.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,325 @@
+        encode_value(output, value)
+    }
+
+    pub fn encode(input: &BTreeMap<String, Value>, fields: &[String]) -> String {
```
I think you could avoid some cloning below if you had `input` just be a `BTreeMap<String, Value>`. We don't need it outside of calling this function.

jszwedko commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,325 @@
+    }
+
+    fn flatten(
+        input: &BTreeMap<String, Value>,
```
This would also change to take ownership of `input` to avoid unneeded cloning below.

This could get a bit tricky, but I think would help improve performance here.

jszwedko commented on docs/reference/remap/functions/encode_logfmt.cue:
```diff
@@ -0,0 +1,56 @@
+            type: ["array"]
+        },
+	]
+	internal_failure_reasons: []
```
```suggestion
	internal_failure_reasons: ["`fields_ordering` contains a non-string element"]
```

I think this is the only way this function could fail?

jszwedko reviewed (commented):
Great work @oktal ! The flattening logic looks great. This actually made me realize we aren't unflattening in `parse_logfmt`. I'll open a separate issue to follow up on that.

I left a few inline comments, but nothing too major. Let me know what you think!

oktal commented on lib/vrl/compiler/src/value/convert.rs:
```diff
@@ -225,8 +225,8 @@ impl Value {
         match self.as_bytes() {
             Some(bytes) => Ok(String::from_utf8_lossy(&bytes)),
             None => Err(Error::Expected {
-                got: Kind::Bytes,
```
Yes, the error message displayed the wrong order in my test.

It was in the right order above https://github.com/timberio/vector/blob/e7379657a86943c1ff1897f3a45ff1314da6e502/lib/vrl/compiler/src/value/convert.rs#L218

jszwedko replied:
Aha, yes, of course 😄 Thanks for making that change!

oktal commented:
Thanks @jszwedko !

I addressed your comments, and modified the code accordingly.
You were right, I removed some `clone()`  with value passing.

I also updated the doc.

StephenWakely commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,324 @@
+use std::result::Result;
+use vrl::prelude::*;
+
+mod logfmt {
```
I don't think it's necessary to have a separate submodule here.

oktal replied:
Yeah, I introduced a submodule so that I could write `logfmt::encode` in the implementation of the function. I think it makes it more pleasant to read, but I can remove it if you think that's not necessary.

jszwedko replied:
We've typically just had private methods at the top level in these modules. I don't feel strongly either way, but did make @StephenWakely 's suggestion in dbd632c

StephenWakely reviewed (changes requested):
Yeah this looks great! If we could just look into reducing those string allocations, I'd be happy to see this go!

oktal commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,324 @@
+            .to_string()
+            .replace(r#"\"#, r#"\\"#)
+            .replace(r#"""#, r#"\""#)
+            .replace("\n", r#"\\n"#);
```
Ah, this is how I implemented it first and that change was suggested

https://github.com/timberio/vector/pull/6985/commits/e7379657a86943c1ff1897f3a45ff1314da6e502

@jszwedko ? :smiley:

jszwedko replied:
😄 I had originally suggested doing it this way but yes, you are right, I neglected this allocates intermediate strings. Your suggested change does make a roughly 50% change to throughput when I tested locally.

StephenWakely replied:
:-) It's not a proper pr if it doesn't go back and forth at least a few times!

oktal commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,324 @@
+    use vrl::prelude::*;
+
+    fn encode_string(output: &mut String, str: &str) {
+        let needs_quotting = str.find(' ').is_some();
```
Good catch, thanks !

jszwedko commented:
Awesome, thanks @oktal ! I made a few more tweaks to reduce allocations along with @StephenWakely's suggestion in https://github.com/timberio/vector/pull/6985#discussion_r626588265 . I think this is good, but will wait for @StephenWakely to give it another look.

StephenWakely commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,305 @@
+            args: func_args![value:
+                btreemap! {
+                    "lvl" => "info",
+                    "msg" => r#"payload: {"code": 200}\n"#
```
Can we add another field to test here to test the other escapes and also test that keys are being escaped:

```rust
"another_field" => "some\nfield\\and things"
```

Should be:

```rust
want: Ok(r#"lvl=info msg="payload: {\"code\": 200}\\n" "another field": "some\\nfield\\and things""#),
```

jszwedko replied:
👍 added in b7aa5c7

oktal commented:
Very cool, thanks.

Just for my Rust' culture, you transformed the `flatten` function a little bit to make it more generic, right ?
Also, what's the performance cost of `Box`'ing an iterator vs moving from a `BTreeMap` ?

jszwedko commented:
> Just for my Rust' culture, you transformed the `flatten` function a little bit to make it more generic, right ?
> Also, what's the performance cost of Box'ing an iterator vs moving from a BTreeMap ?

Yeah, for that I was just trying to avoid allocating a bunch of `BTreeMaps` that we end up dropping very soon after they are created. Instead we just build up an iterator of values that can be collected all at once. It'd be nice to even avoid that last collection, but I couldn't see an easy way to do that with needing to fetch arbitrary values to order them if `field_ordering` is passed.

StephenWakely commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -225,9 +225,10 @@ mod tests {
                     "lvl" => "info",
-                    "msg" => r#"payload: {"code": 200}\n"#
+                    "msg" => r#"payload: {"code": 200}\n"#,
+                    "another_field" => "some\nfield\\and things"
```
What happens if the key has a space in it? This would be possible with Vrl..

jszwedko replied:
It is quoted in the same way that values are quoted. I added another field to test that in 5716e7820

StephenWakely replied:
Perfect! Yeah this is looking good.

JeanMertz commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,307 @@
+    }
+
+    fn type_def(&self, state: &state::Compiler) -> TypeDef {
+        self.value.type_def(state).fallible().bytes()
```
nit:

The output of this function doesn't depend on its input (which you can see since we're setting both the faillibity (`fallible()`) and the type (`bytes()`) manually), so to avoid any confusion, I'd write this as:

```rust
TypeDef::bytes().fallible()
```

jszwedko replied:
6a22ec0

JeanMertz commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,307 @@
+}
+
+fn encode_string(output: &mut String, str: &str) {
+    let needs_quoting = str.find(' ').is_some();
```
Is this check sufficient? Should we instead do?

```rust
str.chars().any(char::is_whitespace)
```

JeanMertz commented on lib/vrl/stdlib/src/encode_logfmt.rs:
```diff
@@ -0,0 +1,307 @@
+            encode_string(output, &val)
+        }
+        _ => {
+            let val = format!("{}", value);
+            encode_string(output, &val)
```
```suggestion
            encode_string(output, &value.to_string())
```

jszwedko replied:
0f99069

jszwedko commented:
Merging this in. Thanks again for all of your work on this @oktal ! I think it's a great addition and also sets the groundwork for using `logfmt` as an `encoding.codec` option on the console sink, for example.
